### PR TITLE
Allow user to compile only some plugins ##build

### PR DESCRIPTION
**Pull request by ret2libc**
**DO NOT SQUASH**

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

* Move plugins definitions to `p/meson.build`
* Move include files meson parts to `librz/include/meson.build`
* Add user options to specify which plugins to compile and how (static vs shared)
* Remove BinLdr plugins because not used
* Add CI job executed on dist/release to compile everything shared.

Apart from allowing the user to customize a lot his own installation for special needs, this will simplify a lot the extraction of some plugins outside of the main repository, as all the headers/files/dependencies are already listed for each plugin.

Only downside I can find of this approach is that `librz/meson.build` has became a bit complicated, in an effort to reduce the duplication. To make things simpler to read I'd have to duplicate all the code used to statically/dynamically compile plugins in each module meson.build.

Please review commit-per-commit, it should be easier.

- [ ] dist checks https://github.com/rizinorg/rizin/pull/839
- [x] make sure pyc plugins are enabled
- [x] test shared plugins on windows (mainly to ensure windows plugins can be compiled on their own)
- [ ] fix debian old compilation

**Test plan**

Try things like:
```
$ meson -Danalysis_plugins=shared.* -Dparse_plugins= -Dasm_plugins=static.* build
$ meson -Dbin_plugins=any,elf,shared.elf64,static.xtr_fatmach0 build2
```

**Closing issues**

...

### CODE REVIEW

**Comment by ret2libc** [comment]
There are some UAF when dealing with shared plugins, but I think I can deal with this immediately after this PR to avoid adding too much stuff. It just happens when loading shared plugins.

EDITED:
The reason I want do it later is that I think the problem was "always" there, but given that shared plugins were never really tested, nobody noticed. Again, apart from the nice-to-have build options this PR provides, I think it is a mandatory step for moving some plugins out of main repo in the future(far or not that is).

**Comment by XVilka** [comment]
Will it need to be redone again when we merge RzAsm and RzAnalysis plugins into RzArch? If yes, maybe it's better to postpone this change?

**Comment by ret2libc** [comment]
> Will it need to be redone again when we merge RzAsm and RzAnalysis plugins into RzArch? If yes, maybe it's better to postpone this change?

Redone? Of course you will need to adjust the meson files if you move files around, but the amount of work is no different/harder/easier if you do it before or after.

**Comment by XVilka** [comment]
```
"cl" "-Ilibrz\libio_plugin_debug.a.p" "-Ilibrz" "-I..\librz" "-I." "-I.." "-Ilibrz\include" "-I..\librz\include" "-I..\shlr\spp" "-Isubprojects\sdb" "-I..\subprojects\sdb" "-Isubprojects\sdb\src" "-I..\subprojects\sdb\src" "-Isubprojects\zlib-1.2.11" "-I..\subprojects\zlib-1.2.11" "/MT" "/nologo" "/showIncludes" "/W2" "/O2" "/Gw" "-DRZ_PLUGIN_INCORE=1" "/Fdlibrz\libio_plugin_debug.a.p\io_p_io_debug.c.pdb" /Folibrz/libio_plugin_debug.a.p/io_p_io_debug.c.obj "/c" ../librz/io/p/io_debug.c
../librz/io/p/io_debug.c(55): fatal error C1083: Cannot open include file: 'w32dbg_wrap.h': No such file or directory
[2/1455] Compiling C object shlr/librzwinkd.a.p/winkd_iob_pipe.c.obj
[3/1455] Compiling C object shlr/librzwinkd.a.p/winkd_transport.c.obj
```
https://github.com/rizinorg/rizin/pull/832/checks?check_run_id=2129338919#step:6:890

File: librz/meson.build [hunk 1]
```diff
@@ -1,460 +1,156 @@
...
+      set_variable('@0@_enabled_static_plugins_str'.format(module), static_plugins_str)
+    else
+      warning('No @0@_plugins variable defined'.format(module))
+    endif
+  endif
+
+  subdir(module)
+
+  if fs.is_dir(module / 'p') and fs.exists(module / 'p' / 'meson.build')
+    # compile static plugins
```
Comment (ret2libc) [inline librz/meson.build:130]: wrong comment

**Comment by XVilka** [comment]
Same here. It's better to postpone.

**Comment by ret2libc** [comment]
Yes I can close this. I will get back to this after we improve the RzLib part.

**Comment by ret2libc** [comment]
Do not delete the branch though, please.

**Pull request closed without merging**